File: Arrays/122-Best-Time-to-Buy-and-Sell-Stock-II/Python/122-Best-Time-to-Buy-and-Sell-Stock-II-Recursion+Memoization.py
"""
Problem: 122. Best Time to Buy and Sell Stock II
Approach: Recursion + Memoization

Store the answer for every (idx, buy) state
to avoid recalculating the same state.

Time Complexity: O(n)
Space Complexity: O(n)
"""


class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        n = len(prices)

        dp = [[-1] * 2 for _ in range(n)]

        def f(idx, buy):
            if idx == n:
                return 0

            if dp[idx][buy] != -1:
                return dp[idx][buy]

            if buy:
                dp[idx][buy] = max(
                    -prices[idx] + f(idx + 1, 0),
                    f(idx + 1, 1)
                )

            else:
                dp[idx][buy] = max(
                    prices[idx] + f(idx + 1, 1),
                    f(idx + 1, 0)
                )

            return dp[idx][buy]

        return f(0, 1)

File: Arrays/122-Best-Time-to-Buy-and-Sell-Stock-II/JavaScript/122-Best-Time-to-Buy-and-Sell-Stock-II-Recursion+Memoization.js
/**
 * Problem: 122. Best Time to Buy and Sell Stock II
 * Approach: Recursion + Memoization
 *
 * Store the answer for every (idx, buy) state
 * to avoid recalculating the same state.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

/**
 * @param {number[]} prices
 * @return {number}
 */

var maxProfit = function(prices) {
    const n = prices.length;

    const dp = Array.from(
        { length: n },
        () => Array(2).fill(-1)
    );

    function f(idx, buy) {
        if (idx === n) {
            return 0;
        }

        if (dp[idx][buy] !== -1) {
            return dp[idx][buy];
        }

        if (buy) {
            dp[idx][buy] = Math.max(
                -prices[idx] + f(idx + 1, 0),
                f(idx + 1, 1)
            );
        } else {
            dp[idx][buy] = Math.max(
                prices[idx] + f(idx + 1, 1),
                f(idx + 1, 0)
            );
        }

        return dp[idx][buy];
    }

    return f(0, 1);
};